### PR TITLE
language picker in docs

### PR DESCRIPTION
**Pull request by jwunderl**
building on https://github.com/microsoft/pxt/pull/5502

One main question; is putting the lang cookie into the `/` path alright? it seems like it would be (just a language choice, and won't conflict with other things on the domain as it's pretty specific (PXT_LANG)), and otherwise it will have really weird behavior when you e.g. change lang in `/courses/csintro` and it doesn't change in /docs or the editor.

~draft as I'm still trying to get the modal focus trap working properly, it wasn't correctly trapping focus (keyboard / screenreader accessibility), haven't got a fix quite yet. Might have to switch up the modal implementation (but that is potentially worth doing anyways, as this one used the jquery modal). Current hack solution in https://github.com/microsoft/pxt/pull/6064/commits/ebdaa2ce65d12a7bb06902b5504c2a678328d904, but it's kinda gross.~ Modal behavior effectively matching the main apps modals, needs proper fix later for both.

### CODE REVIEW

File: docfiles/footer.html [hunk 1]
```diff
@@ -20,12 +20,12 @@
         </div>
         <div class="ui section divider"></div>
         <div class="ui container horizontal small divided link list">
+            <a id="langpicker" tabindex="0" class="ui item" rel="noopener">
+                <i class="icon xicon globe" aria-hidden="true" role="presentation"></i>Language
+            </a>
             <a class="item" href="https://makecode.com/privacy" target="_blank" rel="noopener">Privacy &amp; Cookies</a>
             <a class="item" href="https://makecode.com/termsofuse" target="_blank" rel="noopener">Terms Of Use</a>
             <a class="item" href="https://makecode.com/trademarks" target="_blank" rel="noopener">Trademarks</a>
-            <!-- @ifdef crowdinproject -->
```
Comment (pelikhan) [inline docfiles/footer.html]: why is this removed?

Comment (jwunderl) [reply docfiles/footer.html]: I bumped it to the bottom of the lang selection page to match editor intro, was going to ask you about that on monday though.  Will put it back here if you prefer

File: docfiles/docs.js [hunk 2]
```diff
@@ -280,6 +281,135 @@ function renderSnippets() {
...
+            var currentOption = langOptions[i];
+
+            currentOption.onclick =  function(e) {
+                var langId = e.currentTarget.dataset.lang;
+                if (!pxt.Util.allLanguages[langId]) {
+                    return;
+                }
+                setCookieLang(langId);
+                if (langId !== initialLang) {
+                    pxt.tickEvent("menu.lang.changelang." + langId);
```
Comment (abchatra) [inline docfiles/docs.js]: Fix this in the new format. dynamic langId shouldn't be done here. 
        pxt.tickEvent(`menu.lang.changelang`, { lang: langId });

File: docfiles/docs.js [hunk 3]
```diff
@@ -280,6 +281,135 @@ function renderSnippets() {
...
+                    // In react app before reload we are using pxt.winrt.releaseAllDevicesAsync()
+                    // In docs we currently don't have access to pxt.winrt
+                    location.reload();
+                } else {
+                    pxt.tickEvent("menu.lang.samelang." + langId);
```
Comment (abchatra) [inline docfiles/docs.js]: pxt.tickEvent(`menu.lang.samelang`, { lang: langId, docs: "true" });

File: docfiles/docs.js (hunk 2 above)
Comment (abchatra) [reply docfiles/docs.js]: Also probably add a notion of docs too.
pxt.tickEvent("menu.lang.changelang", { lang: langId, docs: "true" });

File: pxtlib/util.ts [hunk 4]
```diff
@@ -970,6 +970,25 @@ namespace ts.pxtc.Util {
...
+        return cookieValue && cookieValue[1] || null;
+    }
+
+    export function setCookieLang(langId: string, docs = "false") {
+        if (!pxt.Util.allLanguages[langId]) {
+            return;
+        }
+
+        if (langId !== getCookieLang()) {
+            pxt.tickEvent(`menu.lang.setcookielang`, { lang: langId, docs});
```
Comment (abchatra) [inline pxtlib/util.ts]: needs docs : "true"

File: docfiles/docs.js [hunk 5]
```diff
@@ -280,6 +281,108 @@ function renderSnippets() {
...
+    cardEl.setAttribute('role', 'option');
+    cardEl.setAttribute('aria-label', locale.englishName);
+    cardEl.setAttribute('tabindex', '0');
+    cardEl.appendChild(contentEl);
+
+    return cardEl;
+}
+
+function setupLangPicker() {
+    var appTheme = pxt.appTarget.appTheme;
```
Comment (pelikhan) [inline docfiles/docs.js]: check that appTheme exists

File: pxtlib/util.ts [hunk 6]
```diff
@@ -970,6 +970,25 @@ namespace ts.pxtc.Util {
         return false;
     }
 
+    export function getCookieLang() {
+        const cookiePropRegex = new RegExp(`${pxt.Util.escapeForRegex(pxt.Util.pxtLangCookieId)}=(.*?)(?:;|$)`)
+        const cookieValue = cookiePropRegex.exec(document.cookie);
+        return cookieValue && cookieValue[1] || null;
+    }
+
+    export function setCookieLang(langId: string, docs: "true" | "false" = "false") {
```
Comment (pelikhan) [inline pxtlib/util.ts]: "true" | "false" = "false" -> boolean?

File: pxtlib/util.ts [hunk 7]
```diff
@@ -970,6 +970,25 @@ namespace ts.pxtc.Util {
         return false;
     }
 
+    export function getCookieLang() {
```
Comment (pelikhan) [inline pxtlib/util.ts]: since this relies on a browser feature, it should be in pxt.BrowserUtils.

**Comment by jwunderl** [comment]
Hmm, both language pickers are breaking in minecraft at the moment due to the `nb` language not being pxt.Util.allLanguages[code]. Should I add checks for that and just drop them from the language selection or should nb info be in or added to pxt.Util.allLanguages[code] ?

**Comment by pelikhan** [comment]
i would recommend fixing nb issues separately from this.

File: docfiles/docs.js [hunk 8]
```diff
@@ -279,6 +280,110 @@ function renderSnippets() {
     });
 }
 
+function languageOption(code) {
+    var locale = pxt.Util.allLanguages[code];
+    if (!locale)
```
Comment (abchatra) [inline docfiles/docs.js]: Lets remove these check. We should crash instead of silently succeed if the apptheme locale is not a subset of all available locales

File: docfiles/footer.html [hunk 9]
```diff
@@ -20,12 +20,12 @@
...
-            <a id="translatebtn" tabindex="0" class="item">Translate this page</a>
```
Comment (pelikhan) [inline docfiles/footer.html]: please keep this. putting this at the bottom of the language picker is burying too much

Comment (abchatra) [reply docfiles/footer.html]: We talked about this in the design meeting.  Consensus language picker is good place to put this in. Consistent with the editor. If you are translating you will click on the language selection to see.

**Comment by pelikhan** [comment]
@jwunderl  is this ready?
